crypto/x509v3/v3_purp.c

```diff
@@ -781,11 +783,16 @@ int X509_check_issued(X509 *issuer, X509 *subject)
     x509v3_cache_extensions(subject);
 
     if (subject->akid) {
-        int ret = X509_check_akid(issuer, subject->akid);
+        ret = X509_check_akid(issuer, subject->akid);
         if (ret != X509_V_OK)
             return ret;
     }
 
+    CRYPTO_THREAD_read_lock(subject->lock); /* Prevent data races here */
+    CRYPTO_THREAD_read_lock(issuer->lock);
```

Comment:
this looks like a possible deadlock. what if subject==issuer, then you locj the same object twice?

Reply:
correction: _it is_ a dead-lock.

Reply:
AFAIK, all locks that are used are recursive. So locking the same thing twice (in the same thread) is OK (as long as you unlock it the same number of times.
Is that not true for all (ports of) openSSL?

Reply:
The original problem was reported by Inspector. In my original issue #7656 there was an attachment OpenSSLdatarace-5.txt, which is an Intel Inspector report of the problem. If you look at that file, you can see a conflict between a READ being done in RAND_DRBG_bytes (drbg_lib.c:658) and a WRITE in the RAND_DRBG_reseed (drbg_lib.c:441).
Those were reproducible before the fix, gone after the fix.
And if drbg is thread local, that does not necessarily mean drbg->parent is local...
Since the problem was not going to be fixed by someone else, I did what I could to make fix our production-environment crashes, and that worked....

I'll have a look at the travis issues, I just learned what that is 5 minutes ago...

Reply:
> If you look at that file, you can see a conflict between a READ being done in RAND_DRBG_bytes (drbg_lib.c:658) and a WRITE in the RAND_DRBG_reseed (drbg_lib.c:441).

Okay, I see, but that was fixed with #7399 also around the same time.
So first you should repeat your tests with the 1.1.1a version.
It is however possible that tsan_load and tsan_store need an adjustment
for your tool, because as the name implies it is mainly used to prevent
false positive error reports from TSAN.  But I am not aware of atomics
in windows with the weak semantics that we need here.

Reply:
> I did what I could to make fix our production-environment crashes, and that worked....

Note: I the race condition in #7399 did not result in crashes, just a theoretical issue
with too late re-seeding the DRBG slave.  So there must be a different reason
for those crashes you have observed.

Reply:
I'll try and retest with the 1.1.1a. Will take a number of days (at least). More later.
I can also reproduce the hang with the test (see below). Not Good.
I'll make sure I've crossed the T's and dotted the I's before trying again here.... Thank you for the fast response!

Core was generated by `../../test/drbgtest'.
Program terminated with signal 3, Quit.
#0  0x00007ffff79c62ce in pthread_rwlock_wrlock () from /lib64/libpthread.so.0
Missing separate debuginfos, use: debuginfo-install glibc-2.17-260.el7.x86_64
(gdb) bt
#0  0x00007ffff79c62ce in pthread_rwlock_wrlock () from /lib64/libpthread.so.0
#1  0x0000000000426c09 in CRYPTO_THREAD_write_lock ()
#2  0x000000000042a809 in rand_drbg_get_entropy ()
#3  0x000000000040424f in get_entropy_hook ()
#4  0x0000000000428aa2 in RAND_DRBG_instantiate ()
#5  0x0000000000429044 in rand_drbg_restart ()
#6  0x0000000000429176 in RAND_DRBG_generate ()
#7  0x000000000042945a in RAND_DRBG_bytes ()
#8  0x00000000004031d8 in test_drbg_reseed ()
#9  0x000000000040454e in test_rand_drbg_reseed ()
#10 0x00000000004062df in run_tests ()
#11 0x00000000004036b5 in main ()